### PR TITLE
Forcing error in Jenkins builds on exec error/fail

### PR DESCRIPTION
**Pull request by jupierce**
QE requested to fail job if there are errors in the underlying exec API: https://bugzilla.redhat.com/show_bug.cgi?id=1390491

### CODE REVIEW

**Comment by jupierce** [comment]
ptal @gabemontero

**Comment by jupierce** [comment]
The new behavior should be as follows:
- Timeouts/failures/errors should cause traditional builds to fail.
- DSL jobs will only fail because of exec timeouts. This should allow DSL scripts to handle errors via scripting.

**Comment by bparees** [comment]
lgtm but will definitely defer to @gabemontero

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java [hunk 1]
```diff
@@ -130,19 +135,25 @@ public void onFailure(IOException e) {
...
         try {
-            success = latch.await(Math.max(remainingWaitTime, 1), TimeUnit.MILLISECONDS);
+            timeout = ! latch.await(Math.max(remainingWaitTime, 1), TimeUnit.MILLISECONDS);
         } catch (InterruptedException ie) {
         }
 
-        if ( success ) {
-            listener.getLogger().println(String.format(MessageConstants.EXIT_EXEC_GOOD, DISPLAY_NAME));
+        if ( timeout ) {
+            listener.getLogger().println(String.format(MessageConstants.EXIT_EXEC_BAD, DISPLAY_NAME, "Timeout waiting for exec to finish"  ));
```
Comment (gabemontero) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java]: create a message constant for "timeout waiting for exec to finish"

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java [hunk 2]
```diff
@@ -130,19 +135,25 @@ public void onFailure(IOException e) {
...
+            return false;
+        }
+
+        if ( reportError.get() ) {
+            listener.getLogger().println(String.format(MessageConstants.EXIT_EXEC_BAD, DISPLAY_NAME, "Error or failure during exec"));
```
Comment (gabemontero) [inline src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java]: create a message constant for "error or failure during exec"

**Review by gabemontero — COMMENTED** [review]
Couple of minor comments.

Also, the README should be updated to reflect the behavior here, especially given the diff between freestyle and pipeline.

Otherwise, LGTM.  I'm amenable to merging this evening and letting @jupierce handle the comments in a subsequent PR tomorrow if we are anxious to let QE validate the change.  Just let me know - thanks.

**Comment by bparees** [comment]
> I'm amenable to merging this evening and letting @jupierce handle the comments in a subsequent PR tomorrow if we are anxious to let QE validate the change. Just let me know - thanks.

i think we're fine to wait.

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java (hunk 1 above)
Comment (jupierce) [reply src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java]: Will do.

File: src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java (hunk 2 above)
Comment (jupierce) [reply src/main/java/com/openshift/jenkins/plugins/pipeline/model/IOpenShiftExec.java]: Will do.

**Comment by jupierce** [comment]
Additions to README.MD and new string constants in last push. 

Types of errors:
(1) Pod not present
(2) Specified container not present in pod
(3) Command not found  (e.g. "echo1" vs "echo")
(4) Timeout

Freestyle job killed by: 1, 2, 3, 4
DSL job killed by: 1, 5

**Comment by gabemontero** [comment]
IGTM - even though @jupierce 's extended test fixes got in before this started, I need to make the tweaks to the PR test job for everything to take hold.  So not waiting on this PR test job.

**Pull request merged**